winlogbeat/eventlog/eventlog.go

```diff
@@ -138,3 +146,154 @@ func incrementMetric(v *expvar.Map, key interface{}) {
 		v.Add(strconv.Itoa(int(t)), 1)
 	}
 }
+
+// defaultLagPolling is the default polling period for inputMetrics.sourceLagN.
+const defaultLagPolling = time.Minute
+
+// inputMetrics handles event log metric reporting.
+type inputMetrics struct {
+	unregister func()
+	done       chan struct{}
+
+	lastBatch    time.Time
+	lastRecordID *atomic.Uint64
+
+	name        *monitoring.String // name of the provider being read
+	events      *monitoring.Uint   // total number of events received
+	dropped     *monitoring.Uint   // total number of discarded events
+	errors      *monitoring.Uint   // total number of errors
+	batchSize   metrics.Sample     // histogram of the number of events in each non-zero batch
+	sourceLag   metrics.Sample     // histogram of the difference between timestamped event's creation and reading
+	sourceLagN  metrics.Sample     // histogram of difference between the consumer's log offset and the producer's log offset
+	batchPeriod metrics.Sample     // histogram of the elapsed time between non-zero batch reads
+}
+
+// newInputMetrics returns an input metric for windows event logs. If id is empty
+// a nil inputMetric is returned. The ID delta between OS events and read events
+// will be polled each poll duration.
+func newInputMetrics(name, id string, poll time.Duration) *inputMetrics {
+	if id == "" {
+		return nil
+	}
+	reg, unreg := inputmon.NewInputRegistry(name, id, nil)
+	out := &inputMetrics{
+		unregister:  unreg,
+		done:        make(chan struct{}),
+		name:        monitoring.NewString(reg, "provider"),
+		events:      monitoring.NewUint(reg, "received_events_total"),
+		dropped:     monitoring.NewUint(reg, "discarded_events_total"),
+		errors:      monitoring.NewUint(reg, "errors_total"),
+		batchSize:   metrics.NewUniformSample(1024),
+		sourceLag:   metrics.NewUniformSample(1024),
+		batchPeriod: metrics.NewUniformSample(1024),
+	}
+	out.name.Set(name)
+	_ = adapter.NewGoMetrics(reg, "received_events_count", adapter.Accept).
+		Register("histogram", metrics.NewHistogram(out.batchSize))
+	_ = adapter.NewGoMetrics(reg, "source_lag_time", adapter.Accept).
+		Register("histogram", metrics.NewHistogram(out.sourceLag))
+	_ = adapter.NewGoMetrics(reg, "batch_read_period", adapter.Accept).
+		Register("histogram", metrics.NewHistogram(out.batchPeriod))
+
+	if poll > 0 && runtime.GOOS == "windows" {
+		out.lastRecordID = &atomic.Uint64{}
```

Comment:
`out.sourceLagN` is never initialized which will cause a panic when something tries to take a snapshot of the metrics. You can see this if you start Winlogbeat with `logging.metrics.namespaces: [stats, dataset]` and wait 30s.

And I'm thinking that since this metric updates every 1 minute that we need a smaller reservoir size like say 15 so we are looking at samples over the last 15 minutes.

```suggestion
		out.lastRecordID = &atomic.Uint64{}
		out.sourceLagN = metrics.NewUniformSample(15)
```

Reply:
Fixed